File: src/com/family/solitaire/model/Game.java
/*
 * Game.java
 *
 * Created on November 29, 2006, 5:09 PM
 */

package com.family.solitaire.model;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

/**
 * A game in progress: the board, the rule it is played under, and its history.
 * Every change goes through here so it can be checked and undone.
 *
 * @author Aaron Ding
 */
public class Game {

    public Game() {
        board = new Board();
        rule = Rule.EASY;
    }

    /** Deals game number {@code dealNumber}; the same number always deals the same cards. */
    public void deal(Rule rule, int dealNumber) {
        Deck deck = new Deck();
        deck.shuffle(new Random(dealNumber));
        board.deal(deck);
        start(rule, dealNumber);
    }

    void start(Rule rule, int dealNumber) {
        this.rule = rule;
        this.dealNumber = dealNumber;
        done.clear();
        undone.clear();
        elapsedMillis = 0;
    }

    public boolean move(Move move) {
        if (!rule.isValidMove(move, board))
            return false;
        done.add(board.moveCards(move));
        undone.clear();
        return true;
    }

    public boolean useReserve() {
        if (isReserveUsed())
            return false;
        done.add(board.dealReserve());
        undone.clear();
        return true;
    }

    public boolean canUndo() {
        return !done.isEmpty() && rule.mayUndo(done.get(done.size()-1));
    }

    /** Takes back the last step if the rule allows it; returns it, or null. */
    public Step undo() {
        return canUndo() ? undoLast() : null;
    }

    /** Takes back the last step whatever the rule says, for finishing a game the player gave up. */
    public Step forceUndo() {
        return done.isEmpty() ? null : undoLast();
    }

    private Step undoLast() {
        Step step = done.remove(done.size()-1);
        if (step.isReserveDeal())
            board.undealReserve();
        else
            board.unmoveCards(step);
        undone.add(step);
        return step;
    }

    public boolean canRedo() {
        return !undone.isEmpty();
    }

    /** Replays the last undone step; returns it, or null. */
    public Step redo() {
        if (undone.isEmpty())
            return null;
        Step step = undone.remove(undone.size()-1);
        Step again = step.isReserveDeal() ? board.dealReserve() : board.moveCards(step.getMove());
        done.add(again);
        return again;
    }

    // if win return true, otherwise return false
    public boolean checkResult() {
        return rule.checkResult(board);
    }

    public List<Move> getAvailableMoves() {
        return rule.getAvailableMoves(board);
    }

    /** No hint left and the reserve is gone: the original "You Lose!" test. */
    public boolean noMovesLeft() {
        return isReserveUsed() && getAvailableMoves().isEmpty();
    }

    public boolean isReserveUsed() { return board.getReserve().used(); }

    public Board getBoard() { return board; }
    public Column getColumn(int i) { return board.getColumn(i); }
    public Reserve getReserve() { return board.getReserve(); }
    public Rule getRule() { return rule; }
    public String getLevel() { return rule.getLevel(); }

    /** The deal number, or 0 if unknown (a game saved by an older version). */
    public int getDealNumber() { return dealNumber; }

    public List<Step> getHistory() { return Collections.unmodifiableList(done); }
    public int getMoveCount() { return done.size(); }

    public long getElapsedMillis() { return elapsedMillis; }
    public void setElapsedMillis(long millis) { elapsedMillis = millis; }

    void restoreHistory(List<Step> steps) {
        done.clear();
        done.addAll(steps);
        undone.clear();
    }

    private final Board board;
    private Rule rule;
    private int dealNumber;
    private long elapsedMillis;

    private final List<Step> done = new ArrayList<Step>();
    private final List<Step> undone = new ArrayList<Step>();
}

File: web/src/model/game.ts
import { Board, type Move, type Step } from './board';
import { type Rule, hints, isValidMove, isWon } from './rule';

/**
 * A game in progress: the board, the rule it is played under, and its history.
 * Every change goes through here so it can be checked and undone.
 */
export class Game {
  board = new Board();
  rule: Rule;
  /** The deal number, or 0 if unknown. */
  dealNumber = 0;
  elapsedMs = 0;
  private done: Step[] = [];
  private undone: Step[] = [];

  constructor(rule: Rule) {
    this.rule = rule;
  }

  /** Deals game number n; the same number always deals the same cards. */
  static deal(rule: Rule, n: number): Game {
    const game = new Game(rule);
    game.board.deal(n);
    game.dealNumber = n;
    return game;
  }

  move(move: Move): boolean {
    if (!isValidMove(this.rule, move, this.board)) return false;
    this.done.push(this.board.moveCards(move));
    this.undone = [];
    return true;
  }

  useReserve(): boolean {
    if (this.board.reserveUsed) return false;
    this.done.push(this.board.dealReserve());
    this.undone = [];
    return true;
  }

  canUndo(): boolean {
    const last = this.done[this.done.length - 1];
    return !!last && this.rule.mayUndo(last);
  }

  /** Takes back the last step if the rule allows it. */
  undo(): Step | null {
    return this.canUndo() ? this.undoLast() : null;
  }

  /** Takes back the last step whatever the rule says, for finishing a game the player gave up. */
  forceUndo(): Step | null {
    return this.done.length > 0 ? this.undoLast() : null;
  }

  private undoLast(): Step {
    const step = this.done.pop()!;
    if (step.kind === 'reserve') this.board.undealReserve();
    else this.board.unmoveCards(step);
    this.undone.push(step);
    return step;
  }

  canRedo(): boolean {
    return this.undone.length > 0;
  }

  redo(): Step | null {
    const step = this.undone.pop();
    if (!step) return null;
    const again = step.kind === 'reserve' ? this.board.dealReserve() : this.board.moveCards(step.move);
    this.done.push(again);
    return again;
  }

  isWon(): boolean {
    return isWon(this.board);
  }

  hints(): Move[] {
    return hints(this.rule, this.board);
  }

  /** No hint left and the reserve is gone. */
  noMovesLeft(): boolean {
    return this.board.reserveUsed && this.hints().length === 0;
  }

  get history(): readonly Step[] {
    return this.done;
  }

  get moveCount(): number {
    return this.done.length;
  }

  /** Replaces the history, for loading a saved game. */
  restoreHistory(steps: Step[]): void {
    this.done = [...steps];
    this.undone = [];
  }
}
